### ISSUE TITLE
[Bug] mean of gumbelsoftmax output is nan when fp16

### ISSUE THREAD

**Issue by richarddwang** [pytorch/pytorch#41663]
## To Reproduce

```
g = F.gumbel_softmax(torch.randn(128,128,30152, dtype=torch.float16, device='cuda:0').normal_(), dim=-1)
g.mean()
--> tensor(nan, device='cuda:0', dtype=torch.float16)
```

## Expected behavior

when use fp32 with the same scirpt
```
g = F.gumbel_softmax(torch.randn(128,128,30152, dtype=torch.float32, device='cuda:0').normal_(), dim=-1)
g.mean()
--> tensor(3.3165e-05, device='cuda:0')
```
when use fp16 but using another way to do gumbel softmax
```
gumebel_dist = torch.distributions.gumbel.Gumbel(torch.tensor(0., device='cuda:0', dtype=torch.float16), torch.tensor(1., device='cuda:0', dtype=torch.float16))
gg = torch.softmax(torch.randn(128,128,30152, dtype=torch.float16, device='cuda:0').normal_() + gumebel_dist.sample((128,128,30152)), dim=-1)
gg.mean()
--> tensor(3.3140e-05, device='cuda:0', dtype=torch.float16)
```

## My guess
eps or something else in the internal implementation is smaller than the smallest value of fp16

## Environment

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: Quadro GV100
GPU 1: Quadro GV100
GPU 2: Quadro GV100
GPU 3: Quadro GV100

Nvidia driver version: 418.39
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.19.0
[pip] torch==1.5.1
[pip] torchvision==0.6.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] numpy-base                1.18.1           py37hde5b4d6_1  
[conda] pytorch                   1.5.1           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torch                     1.5.1                    pypi_0    pypi
[conda] torchvision               0.6.0                    pypi_0    pypi

**Comment by chenyangh** [comment]
Thank you so much.

**Comment by xvjiarui** [comment]
It may due to the following lines
```
    gumbels = (
        -torch.empty_like(logits, memory_format=torch.legacy_contiguous_format).exponential_().log()
    )  # ~Gumbel(0,1)
```

**Comment by xvjiarui** [comment]
This is still happening in PyTorch 1.8. Any updates?

**Comment by kamalkraj** [comment]
Fixed in 1.9.0

**Comment by richarddwang** [comment]
Also confirmed the fix. Thanks you!

**Comment by chijames** [comment]
I also encounter this problem in a separate project. It turns out that this still happens in version 1.9.0, specifically, the sampled value from exponential_() can still be 0. I don't know how this is fixed...
